File: cpp/src/solutions/palindrome_number.cpp
// LeetCode 9. Palindrome Number in C++

#include "cpp/src/common/common.h"

class Solution {
 public:
  bool isPalindrome(int x) {
    long long y = 0;
    int tmp = x;
    while (tmp > 0) {
      y *= 10;
      y += tmp % 10;
      tmp /= 10;
    }
    return x == y;
  }
};

TEST(isPalindrome, example1) {
  int x = 121;
  bool ans = true;
  EXPECT_EQ(Solution().isPalindrome(x), ans);
}

TEST(isPalindrome, example2) {
  int x = 1121;
  bool ans = false;
  EXPECT_EQ(Solution().isPalindrome(x), ans);
}

TEST(isPalindrome, example3) {
  int x = 10;
  bool ans = false;
  EXPECT_EQ(Solution().isPalindrome(x), ans);
}

TEST(isPalindrome, example4) {
  int x = -101;
  bool ans = false;
  EXPECT_EQ(Solution().isPalindrome(x), ans);
}

File: rust/src/solutions/palindrome_number.rs
// LeetCode 9. Palindrome Number

use super::Solution;

impl Solution {
    pub fn is_palindrome(x: i32) -> bool {
        let mut y = 0;
        let mut tmp = x as i64;
        while tmp > 0 {
            y *= 10;
            y += tmp % 10;
            tmp /= 10;
        }
        x as i64 == y
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let x = 121;
        let ans = true;
        assert_eq!(Solution::is_palindrome(x), ans);
    }

    #[test]
    fn example2() {
        let x = 1121;
        let ans = false;
        assert_eq!(Solution::is_palindrome(x), ans);
    }

    #[test]
    fn example3() {
        let x = 10;
        let ans = false;
        assert_eq!(Solution::is_palindrome(x), ans);
    }

    #[test]
    fn example4() {
        let x = -101;
        let ans = false;
        assert_eq!(Solution::is_palindrome(x), ans)
    }
}
